src/wkt/src/message.rs

```diff
@@ -20,27 +20,52 @@ use crate::AnyError as Error;
 /// A trait that must be implemented by all messages.
 ///
 /// Messages sent to and received from Google Cloud services may be wrapped in
-/// [Any][crate::any::Any]. `Any` uses a `@type` field to encoding the type
+/// [Any][crate::any::Any]. `Any` uses a `@type` field to encode the type
 /// name and then validates extraction and insertion against this type.
 pub trait Message {
     /// The typename of this message.
     fn typename() -> &'static str;
 
+    /// Returns the serializer for this message type
     #[doc(hidden)]
-    /// Store the value into a JSON object.
-    fn to_map(&self) -> Result<Map, Error>
+    fn serializer() -> Box<dyn MessageSerializer<Self>>
     where
-        Self: serde::ser::Serialize + Sized,
+        Self: Sized + serde::ser::Serialize + serde::de::DeserializeOwned + 'static,
     {
-        to_json_object(self)
+        Box::new(DefaultSerializer::<Self>::new())
     }
+}
 
-    #[doc(hidden)]
-    /// Extract the value from a JSON object.
-    fn from_map(map: &Map) -> Result<Self, Error>
-    where
-        Self: serde::de::DeserializeOwned,
-    {
+/// Internal API for message serialization.
+/// This is not intended for direct use by consumers of this crate.
+#[doc(hidden)]
+pub trait MessageSerializer<T> {
+    fn to_map(&self, message: &T) -> Result<Map, Error>;
+    fn from_map(&self, map: &Map) -> Result<T, Error>;
+}
+
+// Default serializer that most types can use
+pub(crate) struct DefaultSerializer<T> {
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> DefaultSerializer<T> {
+    pub fn new() -> Self {
+        Self {
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T> MessageSerializer<T> for DefaultSerializer<T>
```

Comment:
This is good.

I think we want to define another serializer in here:

```rs
// Serializes the type `T` into the `value` field.
pub(crate) struct ValueSerializer<T> {
    _phantom: std::marker::PhantomData<T>,
}

impl<T> ValueSerializer<T> {
    pub fn new() -> Self {
        Self {
            _phantom: std::marker::PhantomData,
        }
    }
}

impl<T> MessageSerializer<T> for ValueSerializer<T>
where
    T: serde::ser::Serialize + serde::de::DeserializeOwned + Sized + Message,
{
    fn to_map(&self, message: &T) -> Result<Map, Error> {
        to_json_other(message)
    }

    fn from_map(&self, map: &Map) -> Result<T, Error> {
        from_other(map)
    }
}
```

Reply:
Done.